File: trabajo 16/radix sort.cpp
#include <iostream>
#include <vector>
#include <cstdlib>   

using namespace std;


void countingSort(vector<int>& arr, int exp) {
    int n = arr.size();
    vector<int> output(n);
    int count[10] = {0};

    // Contar la cantidad de ocurrencias de cada dígito
    for (int i = 0; i < n; i++) {
        int index = (arr[i] / exp) % 10;
        count[index]++;
    }

    // Sumar los conteos acumulados
    for (int i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    // Construir el arreglo de salida (de derecha a izquierda para estabilidad)
    for (int i = n - 1; i >= 0; i--) {
        int index = (arr[i] / exp) % 10;
        output[count[index] - 1] = arr[i];
        count[index]--;
    }

    // Copiar los resultados al arreglo original
    for (int i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}


void radixSort(vector<int>& arr) {
    int maxNum = arr[0];
    for (int num : arr) {
        if (num > maxNum)
            maxNum = num;
    }

    for (int exp = 1; maxNum / exp > 0; exp *= 10) {
        countingSort(arr, exp);
    }
}


int main() {
    srand(time(nullptr));
    vector<int> arr(20);

    // Generar 20 números aleatorios entre 1 y 9999
    for (int i = 0; i < arr.size(); i++) {
        arr[i] = rand() % 9999 + 1;
    }

    cout << "Arreglo desordenado:\n";
    for (int num : arr) cout << num << " ";
    cout << "\n";
    radixSort(arr);

    cout << "\nArreglo ordenado:\n";
    for (int num : arr) cout << num << " ";
    cout << "\n";



    return 0;
}

File: trabajo 16/radix sort.js
function countingSort(arr, exp) {
    const n = arr.length;
    const output = new Array(n).fill(0);
    const count = new Array(10).fill(0);

    // Contar la cantidad de ocurrencias de cada dígito
    for (let i = 0; i < n; i++) {
        const index = Math.floor(arr[i] / exp) % 10;
        count[index]++;
    }

    // Sumar los conteos acumulados
    for (let i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    // Construir el arreglo de salida (de derecha a izquierda para estabilidad)
    for (let i = n - 1; i >= 0; i--) {
        const index = Math.floor(arr[i] / exp) % 10;
        output[count[index] - 1] = arr[i];
        count[index]--;
    }

    // Copiar los resultados al arreglo original
    for (let i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}


function radixSort(arr) {
    let maxNum = Math.max(...arr);

    for (let exp = 1; Math.floor(maxNum / exp) > 0; exp *= 10) {
        countingSort(arr, exp);
    }
}


function main() {
    const arr = Array.from({ length: 20 }, () => Math.floor(Math.random() * 9999) + 1);

    console.log("Arreglo desordenado:");
    console.log(arr.join(", "));

    radixSort(arr);

    console.log("\nArreglo ordenado:");
    console.log(arr.join(", "));
}


main();
